urlbd77b745af_KokaKiwi_MCLauncher_tgz-pJ8-net_contrapunctus_lzma_BenchmarkJ8/src/com/kokakiwi/mclauncher/api/LauncherAPI.java

```diff
@@ -0,0 +1,335 @@
+        }
+        
+        if (dir != null && !dir.exists() && !new File(dir, "bin").mkdirs())
+        {
+            throw new RuntimeException(
+                    "The working directory could not be created: " + dir);
+        }
+        
+        return dir;
+    }
+    
+    public String postUrl(String dest, Map<String, String> params)
+    {
+        return postUrl(dest, params, null);
+    }
+    
+    public String postUrl(String dest, Map<String, String> params,
+            String keyFile)
+    {
+        String result = null;
+        HttpURLConnection connection = null;
+        
+        try
+        {
+            final URL url = new URL(dest);
+            connection = (HttpURLConnection) url.openConnection();
+            
+            connection.setRequestMethod("POST");
+            connection.setRequestProperty("Content-Type",
+                    "application/x-www-form-urlencoded");
+            
+            final StringBuffer sb = new StringBuffer();
+            final Iterator<Entry<String, String>> iterator = params.entrySet()
+                    .iterator();
+            while (iterator.hasNext())
+            {
+                final Entry<String, String> param = iterator.next();
+                
+                sb.append(param.getKey());
+                sb.append('=');
+                sb.append(param.getValue());
+                
+                if (iterator.hasNext())
+                {
+                    sb.append('&');
+                }
+            }
+            
+            connection.setRequestProperty("Content-Length",
+                    Integer.toString(sb.toString().getBytes().length));
+            connection.setRequestProperty("Content-Language", "en-US");
+            
+            connection.setUseCaches(false);
+            connection.setDoInput(true);
+            connection.setDoOutput(true);
+            
+            connection.connect();
+            
+            if (url.getProtocol().contains("https") && keyFile != null)
+            {
+                final Certificate[] certs = ((HttpsURLConnection) connection)
+                        .getServerCertificates();
+                
+                final byte[] bytes = new byte[294];
+                final DataInputStream dis = new DataInputStream(
+                        LauncherAPI.class
+                                .getResourceAsStream("keys/" + keyFile));
+                dis.readFully(bytes);
+                dis.close();
+                
+                final Certificate c = certs[0];
+                final PublicKey pk = c.getPublicKey();
+                final byte[] data = pk.getEncoded();
+                
+                for (int i = 0; i < data.length; i++)
+                {
+                    if (data[i] == bytes[i])
+                    {
+                        continue;
+                    }
+                    throw new RuntimeException("Public key mismatch");
+                }
+            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 201:225, 283:307.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.